File: pycbc/matchedfilter/clayer/cpu/matchedfiltercpu.c
// Copyright (C) 2011 Badri Krishnan, Karsten Wiesner
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


//
// =============================================================================
//
//                                   Preamble
//
// =============================================================================
//
// matchedfiltercpu  

#include <stdio.h>
#include "datavectorcpu.h"
#include "pycbccpu.h"
#include "fftw.h" 
#include "matchedfiltercpu_private.h"
#include "matchedfiltercpu.h"

matched_filter_cpu_t* new_matched_filter_cpu_t(cpu_context_t* context)
{

    matched_filter_cpu_t* c;
    c = (matched_filter_cpu_t*) malloc( sizeof(matched_filter_cpu_t) );

    return c;
}


void delete_matched_filter_cpu_t( matched_filter_cpu_t* p )
{
    free( p );
}


void gen_snr_cpu( cpu_context_t* context,
		          real_vector_single_cpu_t* snr,
		          complex_vector_single_cpu_t* stilde, 
		          complex_vector_single_cpu_t* htilde,
		          complex_vector_single_cpu_t* q,
		          complex_vector_single_cpu_t* qtilde,
		          fftw_complex_single_plan* plan,
		          double f_min,
		          double sigma_sq)
{    

    static unsigned cnt=0;
    int j;
    const int N = 2 * (stilde->meta_data.vector_length - 1);
    double norm = 4.0 / ((double) N * (double) N * stilde->meta_data.delta_x); 
    norm *= norm;
    
    printf("%d: called gen_snr_cpu with context: %p, snr: %p s: %p h: %p q: %p qtilde %p fft_plan: %p, fmin: %f, sigma-sq: %f\n", 
           cnt++, context, snr, stilde, htilde, q, qtilde, plan, f_min, sigma_sq);
    
    /* perform the correlation */
    correlate_complex_freq_vectors( qtilde, stilde, htilde, f_min );

    /* execute complex single ifft (output, input, plan) */
    execute_complex_single_fft(q, qtilde, plan);

    /* normalize the snr */
    for ( j = 0; j < q->meta_data.vector_length; ++j )
      snr->data[j] = (norm/sigma_sq) * 
        (__real__ q->data[j] * __real__ q->data[j] + 
         __imag__ q->data[j] * __imag__ q->data[j]);
  
    snr->meta_data.delta_x = q->meta_data.delta_x;
  
    return;
}

void find_max_cpu( cpu_context_t* context,
                   double* max,
                   unsigned long* index,
                   real_vector_single_cpu_t* snr)
{    
    unsigned long i;

    *max = 0;
    for (i=0; i < snr->meta_data.vector_length; i++) {
        if (snr->data[i] > *max) {
            *max = snr->data[i];
            *index = i;
        }	 
    }
    
    return;
}


/* freq domain correlation of two vectors (suitably whitened) */
void correlate_complex_freq_vectors( complex_vector_single_cpu_t* out,
				     complex_vector_single_cpu_t* x, 
				     complex_vector_single_cpu_t* y, 
				     double f_min)
{
    int k;
    const double df = x->meta_data.delta_x;
    const int bin_min = f_min / df > 1 ? f_min / df : 1;
    const int len = x->meta_data.vector_length;

    for ( k = bin_min; k < len-1; k++ )
    {
        __real__ out->data[k] = __real__ x->data[k] * __real__ y->data[k] - 
                                __imag__ x->data[k] * 
                                ( 0.0 - __imag__ y->data[k] );
        __imag__ out->data[k] = __real__ x->data[k] * 
                                (0.0 - __imag__ y->data[k]) + 
                                __imag__ x->data[k] * __real__ y->data[k];
    }

    out->meta_data.delta_x = df;

    return;
}

File: pycbc/matchedfilter/matchedfilter_cpu.py
# Copyright (C) 2011 Karsten Wiesner
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


#
# =============================================================================
#
#                                   Preamble
#
# =============================================================================
#

"""
MatchedFilter Cpu implementation class for the pycbc package
"""

from matchedfilter_base import MatchedFilterBase
from matchedfilter_base import GenSnrImplementationBase
from matchedfilter_base import MaxImplementationBase

# import processing functions from the clayer 
from matchedfiltercpu import gen_snr_cpu

# import member datavectors
from pycbc.datavector.datavectorcpu import complex_vector_single_t
from pycbc.datavector.datavectorcpu import real_vector_single_t

# for testing data_in
from pycbc.datavector.datavectoropencl import real_vector_single_t as alien_datavector_t

from pycbc.pycbc import CpuProcessingObj

# import FFT members

# Josh to fix the fftwf fftw issue when the setup.py system allows to link to fftw
#from pycbc.fft.fftcpu import fft_real_single_plan
#from pycbc.fft.fftcpu import execute_real_single_reverse_fft_cpu

import logging

class MatchedFilterCpu(MatchedFilterBase, CpuProcessingObj):

    def __init__(self, length=0, delta_x=1):

        self.__logger= logging.getLogger('pycbc.MatchedFilterCpu')
        self.__logger.debug("instanciated MatchedFilterCpu")

        self.__length= length
        self.__delta_x= delta_x
        
        super(MatchedFilterCpu, self).__init__(self.__length, self.__delta_x,
                               GenSnrImplementationCpu, MaxImplementationCpu, 
                               snr_vector_t=    real_vector_single_t, 
                               qtilde_vector_t= complex_vector_single_t, 
                               q_vector_t =     complex_vector_single_t, 
                               derived_mfilt =  self)


class  GenSnrImplementationCpu(GenSnrImplementationBase):

    def __init__(self, owner_mfilt):

        self.__logger= logging.getLogger('pycbc.GenSnrImplementationCpu')
        self.__logger.debug("instanciated GenSnrImplementationCpu")

        super(GenSnrImplementationCpu, self).__init__(owner_mfilt)
    
    def generate_snr(self, context, snr, stilde, htilde):
        """
        Process matched filtering by generating snr timeseries \rho(t)
        """
        
        # for testing data_in()
        self._aliendatavector = alien_datavector_t(4096, 1.0)
        
        
        self._owner_mfilt.data_in(snr)
        self._owner_mfilt.data_in(stilde)
        self._owner_mfilt.data_in(htilde)
        
        self._owner_mfilt.data_in(self._aliendatavector)
            
        # testing done in dadta_in    
        #assert repr(stilde).find("datavectorcpu") >= 0, "try to call gen_snr_cpu() with wrong type of datavector for stilde"
        #assert repr(htilde).find("datavectorcpu") >= 0, "try to call gen_snr_cpu() with wrong type of datavector for htilde"
        #assert repr(snr).find("datavectorcpu") >= 0, "try to call gen_snr_cpu() with wrong type of datavector for snr"

        gen_snr_cpu(context, snr, stilde, htilde, self._owner_mfilt._q, 
                    self._owner_mfilt._qtilde, 1.0, 1.0) # just for prototyping
                                                         # f_min and sigma_sq
                                                         # are magic numbers


class  MaxImplementationCpu(MaxImplementationBase):

    def __init__(self, owner_mfilt):

        self.__logger= logging.getLogger('pycbc.MaxImplementationCpu')
        self.__logger.debug("instanciated MaxImplementationCpu") 

        super(MaxImplementationCpu, self).__init__(owner_mfilt)
    
    def max(self, snr):
        """
        Find the maximum in the generated snr timeseries \rho(t)
        """
        assert repr(snr).find("datavectorcpu") >= 0, "try to call gen_snr_cpu() with wrong type of datavector for snr"
        
        self.__logger.debug("finding maximum of snr series - to be implemented in clayer")            

        return 5.5


